### ISSUE TITLE
Identifying anonymous contact via PATCH deletes existing fields

### ISSUE THREAD

**Issue by pjeby** [mautic/mautic#7558]
## Bug Description
When an anonymous contact is identified via a PATCH request, and it is a duplicate of an already-identified contact, the resulting merged contact lacks any fields that were not included in the PATCH, losing previously-acquired data for the contact.

| Q   | A
| --- | ---
| Mautic version | 2.15.1
| PHP version | 7.1
| Browser | N/A

### Steps to reproduce
1. Create a new anonymous contact via curl get to /mtc
2. With curl, PATCH the contact to include first name, last name, or other fields, as well as email
3. Create a second anonymous contact via curl get /mtc
4. With curl, PATCH the second contact with the same email used in step 2

Result: the contacts are merged, but the field values created in step 2 will be missing.
 
### Log errors

There are no errors; here is the debug info however.  The firstname and lastname shown are the ones added in step 2, which means that it's not the *merge* algorithm at fault, but the subsequent editing action:

```
[2019-05-24 01:45:50] mautic.DEBUG: LEAD: Existing contact ID# 32 found through query identifiers.
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: ID# 34 will be merged into ID# 32
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: Associating 32 with IP XXX.XXX.XXX.XXX
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: Updated firstname from Nyet to Nyet for 32
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: Updated lastname from Russo to Russo for 32
[2019-05-24 01:45:50] mautic.INFO: CONTACT: email is not mergeable for 32 -
[2019-05-24 01:45:50] mautic.INFO: CONTACT: preferred_locale is not mergeable for 32 -
[2019-05-24 01:45:50] mautic.INFO: CONTACT: timezone is not mergeable for 32 -
[2019-05-24 01:45:50] mautic.INFO: CONTACT: attribution_date is not mergeable for 32 -
[2019-05-24 01:45:50] mautic.INFO: CONTACT: attribution is not mergeable for 32 -
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: Adding 0 points from contact ID #34 to contact ID #32 with 0 points
[2019-05-24 01:45:50] mautic.DEBUG: CONTACT: Contact currently does not have any tags
```

The most likely culprit is the logic for populating fields for the patch operation in the case of a duplicate contact.  (Perhaps the fields are first being populated as blanks from the anonymous contact, then applied to the merged contact?)

(Note: issue #7544 *may* be related, but probably **isn't**.  The audit log screenshots there show blanks for the "New Value" of the fields, but in the scenario described here, the audit log does not show any blanks; only the working non-blank changes from step 2 are shown.)

**Comment by pjeby** [comment]
I tried to work around this by creating an additional unique ID, patching the second contact with a new unique value, and then patching the first contact to have the same unique ID, forcing them to be merged.  Unfortunately, this resulted in the invalidation of the old session information for the second contact, which rendered the whole thing useless for my use case.

So next, I tried making a patch to fix the underlying issue here (i.e. PATCH being broken when contacts merge), but that *also* invalidates the old session information for the second contact!

Basically, there's a second issue here: when contacts are merged, their session data is invalidated, because device records are not transferred upon merge.  I've opened a new issue for this, #7559, and submitted my patch for this fix as #7560.

**Comment by escopecz** [comment]
Try to test https://github.com/mautic/mautic/pull/7717 if it fixes this issue, please.

**Comment by pjeby** [comment]
@escopecz: #7560 is already a tested fix for the actual issue, i.e., the application of PATCH values to the wrong contact when merging.

#7717 on the other hand, ISTM would introduce a new bug: that it would be impossible to set a default value via PATCH when a merge is involved.

**Comment by escopecz** [comment]
Thanks for the feedback. Please comment on that PR if you have concerns about it. It may be overlooked here when the PR will be considered to be merged.

**Comment by npracht** [comment]
merged
